portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -273,6 +300,69 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     response.sendRedirect(REDIRECT_URL);
   }
 
+  /** 
+   * Returns a URL that points to the uploaded file, or null if the user didn't upload a file.
+   *
+   * @param request the request sent to this servlet which contains all user submitted data.
+   * @param formInputElementName the name of the file input element on the HTML submission form.
+   * @return URL of the uploaded file hosted on Blobstore.
+   */
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // Check if the file is a valid image.
```

Comment:
You can remove this comment because the function isValidImage() is self explanatory